File: prolog/triangle/triangle.pl
inequality(X, Y, Z) :- Y + Z >= X,
                        X + Z >= Y,
                        X + Y >= Z.

positive(X, Y, Z) :- X > 0, Y > 0, Z > 0.
triangle(X, Y, Z) :- positive(X, Y, Z),
                        inequality(X, Y, Z).

triangle(X, X, X, "equilateral") :- triangle(X, X, X),
                                        !.

triangle(X, Y, Z, "isosceles") :- (X =:= Y; X =:= Z; Y =:= Z),
                                    triangle(X, Y, Z),
                                    !.

triangle(X, Y, Z, "scalene") :- X \= Y, Y \= Z, X \= Z,
                                    triangle(X, Y, Z).

File: cpp/triangle/triangle.cpp
#include "triangle.h"
#include <stdexcept>

namespace triangle {

	const flavor kind(double a, double b, double c)
	{
		if (!isLegalTriangle(a, b, c))
		{
			throw std::domain_error("Not a valid triangle.");
		}

		// For the test data, == on floating point is sufficient.
		// Without knowing more about the input data, there's not much
		// more we can do.
		if (a == b && b == c)
		{
			return flavor::equilateral;
		}

		if (a == b || a == c || b == c)
		{
			return flavor::isosceles;
		}

		return flavor::scalene;
	}

	const bool isLegalTriangle(double a, double b, double c)
	{
		return a > 0 && b > 0 && c > 0
			&& (a + b >= c)
			&& (a + c >= b)
			&& (b + c >= a);
	}

}  // namespace triangle
